### PR TITLE
improve http parse url performance for #path #query #params

### PR DESCRIPTION
**Pull request by qxo**
less gc (reduce create unncessary object) for #params
(if queryString == null, no need QueryStringDecoder)

no need repeat parse path for #query and #params

Signed-off-by: qxo <qxodream@gmail.com>

### CODE REVIEW

File: src/main/java/io/vertx/core/http/impl/HttpUtils.java [hunk 1]
```diff
@@ -424,7 +424,15 @@ public static URI resolveURIReference(URI base, String ref) throws URISyntaxExce
...
-  static String parsePath(String uri) {
+  static String parsePath(final String uri) {
+    final Object o = parsePathAndQueryStartIf(uri);
+    return (String) ( o instanceof Object[] ? ((Object[])o)[0] : o );
+  }
+
+  /**
+   * Extract the path out of the uri and queryStart
+   */
+  static Object parsePathAndQueryStartIf(final String uri) {
```
Comment (pmlopes) [inline src/main/java/io/vertx/core/http/impl/HttpUtils.java]: I think this is a bad name, I can't tell what this method does from it's name

File: src/main/java/io/vertx/core/http/impl/Http2ServerRequestImpl.java [hunk 2]
```diff
@@ -338,20 +342,37 @@ public String uri() {
...
 
   @Override
   public String query() {
-    synchronized (conn) {
-      this.query = uri != null ? HttpUtils.parseQuery(uri) : null;
-      return query;
+    if (query == null) {
+      synchronized (conn) {
+        if (query == null) {
+          query = path() == uri || queryStart == -1 ? null : queryStart > 0 && uri.length() > queryStart
```
Comment (pmlopes) [inline src/main/java/io/vertx/core/http/impl/Http2ServerRequestImpl.java]: This becomes unreadable with so many chain of logic operations and ternary blocks. I believe proper if statements would be preferred and more readable

File: src/main/java/io/vertx/core/http/impl/HttpUtils.java (hunk 1 above)
Comment (qxo) [reply src/main/java/io/vertx/core/http/impl/HttpUtils.java]: Any suggesion  for name ?

File: src/main/java/io/vertx/core/http/impl/Http2ServerRequestImpl.java (hunk 2 above)
Comment (qxo) [reply src/main/java/io/vertx/core/http/impl/Http2ServerRequestImpl.java]: if path already init( path != null ), should we "synchronized (conn) " for #path ?

File: src/main/java/io/vertx/core/http/impl/Http2ServerRequestImpl.java [hunk 3]
```diff
@@ -414,9 +409,7 @@ public String getHeader(CharSequence headerName) {
   public MultiMap params() {
```
Comment (vietj) [inline src/main/java/io/vertx/core/http/impl/Http2ServerRequestImpl.java]: synchronized (conn) should be wrap the whole block we don't really need double checked locking here

**Comment by vietj** [comment]
what's the point of this PR ? has this been spotted as a performance issue in an application ? I don't think this kind of code (called once per request) is on the hot path of an application. For sure it might be faster but it makes code harder to reason about for a performance gain that is not proven to be useful

**Comment by qxo** [comment]
To be a bottleneck , or not ? -- It's depends on  concurrency(c)
if just one request, sure it's useless
But if concurrency over 400, it's pr will have some improve ...
I hope vert.x  will be extreme fast :)

**Comment by pmlopes** [comment]
> But if concurrency over 400, it's pr will have some improve ...

I understand the idea but I find it a bit hard to imagine that in an application the methods: `path()` `query()` or `params()` get called 400 times. Do you have examples of these cases?

**Comment by vietj** [comment]
@qxo this method is usually called from event loop and biased locking is used in practice. What I mean is that we have no evidence that this is an issue in actual applications. In your PR we are trading simplicity for potential performance. Unless it is proven to be a real problem we don't want to make this trade

**Pull request closed without merging**